### PR TITLE
BUG: DataFrame.quantile with NaNs (GH14357)

### LINKED ISSUE #14357: ValueError when trying to compute Quantile
``` python
In [7]: df = pd.DataFrame(np.random.randn(10, 2))

In [8]: df.iloc[1, 1] = np.nan

In [9]: df.quantile(.5)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-66d518aa86c6> in <module>()
----> 1 df.quantile(.5)

/Users/tom.augspurger/Envs/py3/lib/python3.5/site-packages/pandas-0.19.0rc1+21.ge596cbf-py3.5-macosx-10.11-x86_64.egg/pandas/core/frame.py in quantile(self, q, axis, numeric_only, interpolation)
   5152                                      axis=1,
   5153                                      interpolation=interpolation,
-> 5154                                      transposed=is_transposed)
   5155
   5156         if result.ndim == 2:

/Users/tom.augspurger/Envs/py3/lib/python3.5/site-packages/pandas-0.19.0rc1+21.ge596cbf-py3.5-macosx-10.11-x86_64.egg/pandas/core/internals.py in quantile(self, **kwargs)
   3142
   3143     def quantile(self, **kwargs):
-> 3144         return self.reduction('quantile', **kwargs)
   3145
   3146     def setitem(self, **kwargs):

/Users/tom.augspurger/Envs/py3/lib/python3.5/site-packages/pandas-0.19.0rc1+21.ge596cbf-py3.5-macosx-10.11-x86_64.egg/pandas/core/internals.py in reduction(self, f, axis, consolidate, transposed, **kwargs)
   3071         for b in self.blocks:
   3072             kwargs['mgr'] = self
-> 3073             axe, block = getattr(b, f)(axis=axis, **kwargs)
   3074
   3075             axes.append(axe)

/Users/tom.augspurger/Envs/py3/lib/python3.5/site-packages/pandas-0.19.0rc1+21.ge596cbf-py3.5-macosx-10.11-x86_64.egg/pandas/core/internals.py in quantile(self, qs, interpolation, axis, mgr)
   1325             values = _block_shape(values[~mask], ndim=self.ndim)
   1326             if self.ndim > 1:
-> 1327                 values = values.reshape(result_shape)
   1328
   1329         from pandas import Float64Index

ValueError: total size of new array must be unchanged
```

original post follows
---

I have a simple dataframe that I created as follows:

`df[df['Week of'] == week]`

where `week` is a week name I'm filtering by

I have been taking the quartile values of this dataframe as follows:

`df[df['Week of'] == week].quantile(.25)`

However since the update to Pandas 0.19 I am receiving the error (this code worked fine before):

`values = values.reshape(result_shape)`
`ValueError: total size of new array must be unchanged`

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
Closes #14357

I first used `nanpercentile` when available (np >= 1.9), but this only can handle numeric data, not datetimes. So therefore included a custom `_nanpercentile`

I also added some tests regarding empty dataframes/series in comments, as these are still failing (but are also failing on 0.19.0)

### CODE REVIEW

File: pandas/core/internals.py [hunk 1]
```diff
@@ -1315,16 +1314,38 @@ def quantile(self, qs, interpolation='linear', axis=0, mgr=None):
...
 
-            # even though this could be a 2-d mask it appears
-            # as a 1-d result
-            mask = mask.reshape(values.shape)
-            result_shape = tuple([values.shape[0]] + [-1] * (self.ndim - 1))
-            values = _block_shape(values[~mask], ndim=self.ndim)
-            if self.ndim > 1:
-                values = values.reshape(result_shape)
+        def _nanpercentile1D(values, mask, q, **kw):
+            values = values[~mask]
```
Comment (jreback) [inline pandas/core/internals.py]: from pandas.types.common import is_scalar

File: pandas/core/internals.py [hunk 2]
```diff
@@ -1315,16 +1314,38 @@ def quantile(self, qs, interpolation='linear', axis=0, mgr=None):
...
+
+            if len(values) == 0:
+                if lib.isscalar(q):
+                    return self._na_value
+                else:
+                    return np.array([self._na_value] * len(q),
+                                    dtype=values.dtype)
+
```
Comment (jreback) [inline pandas/core/internals.py]: I might move some of this to  to ``pandas.core.nanops`` (though  you might have to move slightly more as that takes axis arg). Its esentially what you are doing here, but in a slightly more general framework. call it ``nanquantile`` (or ``nanpercentile``)

Comment (jreback) [inline pandas/core/internals.py]: I meant move ALL of this; the nanops do everything (based on dtype), are basically ufuncs per-dtype. Its ok for now if you want to merge (to fix the bug). But let's open a new issue to move this code. All of the rest of it is there (for other ops). We don't *do* very much inside the block managers, mainly just assemble blocks, actual calculations are pushed to other routines (numpy or pandas)

Comment (jorisvandenbossche) [inline pandas/core/internals.py]: sounds good, will open a new issue (and one for the failing empty ones as well)

**Pull request merged**